Changing the web team goals to encompass design, extensions, and code insights (#1950)
sourcegraph/about: handbook/engineering/web/goals.md

### Long term

> **_The web team has three areas of focus: creating and maintaining a highly usable and intentionally designed webapp interface; delivering the full, unique value of [extensions](https://docs.sourcegraph.com/extensions) to our users; and developing code insights into an entirely new featureset._**

Reviewer: If we're going to move away from a common theme for the web team and instead list our focus areas explicitly (I'm on board with that) then I think the list would be: the web app, the extensions platform, code host integrations, and code insights. Or the alternative is the way it's described in the [web team page](https://github.com/sourcegraph/about/blob/main/handbook/engineering/web/index.md#growth-plan) where we split it into "web app" and "extensions and integrations".

```suggestion
**_The web team has four areas of focus: creating and maintaining a highly usable and intentionally designed webapp interface; building and growing the [extensions platform](https://docs.sourcegraph.com/extensions); maintaining and expanding code host integrations; and developing code insights into an entirely new featureset._**
```

Reviewer 2: If we look at the [current proposed org structure for 2021](https://about.sourcegraph.com/handbook/engineering/2021_org), it seems the web team could split into 3 distinct teams:
1. Extensibility Team
1. Code Host Integrations Team
1. Web Infrastructure Team

The one area not covered in the org plan is Code Insights.

Considering the org plan has code host integrations and extensibility (which I read as extensions platform) as separate teams I think it makes sense to go with Marek's update of having 4 categories.

Reviewer 3: @jeanduplessis I agree, but note that the org plan has not been updated since it was created 4 months ago. Imo the org plan shouldn't influence our goals, rather our goals should influence the org plan.

Reviewer 2: @felixfbecker thanks for the context and agree that the org plan should follow the goals.

I think Web Infrastructure and Code Insights are easy to rationalize as distinct categories.

The question in my mind then becomes whether native code host integrations (incl the browser extension) should be group as part of the extension platform or as a separate category? 

@felixfbecker do you have any thoughts on this?

Reviewer 2: Asked differently, given an unlimited ability to split the team, would it makes sense to have a team dedicated to the browser extension and native code host integrations, or would it still make sense to rather group it with the extensions platform work?

Reviewer 4: > the org plan shouldn't influence our goals, rather our goals should influence the org plan.

Correct. The next decision point is "when the web team is too big for one manager, how will we split it?"

> **_The web team has three areas of focus: creating and maintaining a highly usable and intentionally designed webapp interface; delivering the full, unique value of [extensions](https://docs.sourcegraph.com/extensions) to our users; and developing code insights into an entirely new featureset._**

Reviewer 3: Thought: Having three focuses is kind of a contradictory statement. Can one be focused on three things at once? This is not necessarily a problem with the wording, but may just be revealing something inherent to these goals. The goal of our previous goal was to have a single focus that other concerns can be folded into as horizontal concerns. We're giving that up, and that may be fine if we can actually deliver on all these things.

Author: I agree that we're giving that up, but I think it's too constricting to try and limit the web team goals to a single overarching narrative that's still specific enough to be meaningful. ("Make a user enjoy using and extending sourcegraph" encompasses all these things but could apply to many teams and tells you nothing – as soon as you get specific enough, it becomes a list of multiple goals.) Good to call this out, but I think it's a fact of our size and responsibility as a single team that we have multiple goals at the moment.

> **Outcome**: Our webapp, browser extensions and native integrations, and code insights all provide our users unique value. These platforms are well-maintained, well-designed, consistent, easy to setup, well-documented, well-tested, performant, and leave users impressed. The webapp helps users discover and employ the full power of Sourcegraph effectively, with a high quality and highly usable interface. The extensions and extensions API provide powerful capabilities to users and a great experience for extension developers. Code insights expose the value of code metadata to users at all levels of an organization. 

Reviewer 2: ```suggestion
**Outcome**: 
1. Our webapp, browser extensions and native integrations, and code insights all provide our users unique value. 
1. These platforms are well-maintained, well-designed, consistent, easy to setup, well-documented, well-tested, performant, and leave users impressed. 
1. The webapp helps users discover and employ the full power of Sourcegraph effectively, with a high quality and highly usable interface. 
1. The extensions and extensions API provide powerful capabilities to users and a great experience for extension developers. 
1. Code insights expose the value of code metadata to users at all levels of an organization. 
```

**suggestion (non-blocking)**: Propose we make this a list so that it's easier to digest.

> **Outcome**: Our webapp, browser extensions and native integrations, and code insights all provide our users unique value. These platforms are well-maintained, well-designed, consistent, easy to setup, well-documented, well-tested, performant, and leave users impressed. The webapp helps users discover and employ the full power of Sourcegraph effectively, with a high quality and highly usable interface. The extensions and extensions API provide powerful capabilities to users and a great experience for extension developers. Code insights expose the value of code metadata to users at all levels of an organization. 

Author: (Updated by removing / changing these for now.)

**Outcome**: Our webapp, browser extensions and native integrations are great platforms to provide our users unique value through extensions. These platforms are well-maintained, consistent, easy to setup, well-documented, well-tested, performant, and show their power through convincing extensions built on top of them. The extensions platform and API provide powerful capabilities to extension developers and a great developer experience.
**Outcome**: Our webapp, browser extensions and native integrations, and code insights all provide our users unique value. These platforms are well-maintained, well-designed, consistent, easy to setup, well-documented, well-tested, performant, and leave users impressed. The webapp helps users discover and employ the full power of Sourcegraph effectively, with a high quality and highly usable interface. The extensions and extensions API provide powerful capabilities to users and a great experience for extension developers. Code insights expose the value of code metadata to users at all levels of an organization. 

### Medium term

To reach our long-term goal, we set the following medium-term goals to guide our short-term iteration plans.

> We will tackle these medium-term goals in order, though expect to have some work done in parallel as we progress.

Author: I think one thing we might consider is adding a "goals completed" like the distribution team [has](https://about.sourcegraph.com/handbook/engineering/distribution/goals_completed). I'm not done editing the specifics of these goals yet and will move some out to a completed goals section/page. (That said, some of these are intentionally left in – the extensions registry is not "done," just in a good place, and I'm keeping a background list of new improvements for it to be done some day tbd.)

> 1. **Make the Sourcegraph web interfaces more consistent and improve discoverability of Sourcegraph features.**

Reviewer 2: **suggestion (non-blocking)**: Can we consider converting this section into a nested numbered list so it's easy to refer to a goal/sub-goal

1. Goal
    1. Sub-goal
    1. Sub-goal

> 1. **Make the Sourcegraph web interfaces more consistent and improve discoverability of Sourcegraph features.**

Author: I'm still playing around with the wording on this one, but good note and note taken.

   With areas like the repository page, user settings area, navigation, settings, command palette UI and overall design affected, we want to provide the best possible UX for using Sourcegraph. 
2. **Make the browser extension and Sourcegraph extensions a core part of Sourcegraph users' experiences.**
   Our extension platform includes the workflow around creating, installing and using extensions, the API exposed to developers and its documentation.
   To grow adoption of extensions, these need to be solid and obviously useful. The extensions action bar needs to be clear and scalable, and help users discover our extensions. We want to track anonymized, general usage of extensions to determine which extensions are most successful at adding value for our users. 

>    We have a powerful extension API, but we know of use cases and customer requests that could put adddional endpoints to use. 

Reviewer 3: These two lines feel very low-level (mentioning very specific projects, and restricting extension improvements to additional endpoints). When goals are too low-level, they can't be used to inform what projects _not_ mentioned here we should take on. Can we open this up a bit again and talk more about the end goal as opposed to the means to get there? In the previous text I simply used the wording "This includes ..." to give examples while not implying that we'll do that and _only_ that.

Author: Updated, and continuing to work on this section tomorrow.

### Long term

**_The web team has four areas of focus:_**

  1. **_Creating and maintaining a highly usable and intentionally designed webapp interface_**
  1. **_Delivering the full, unique value of Sourcegraph [extensions](https://docs.sourcegraph.com/extensions)_**

>   1. **_Maintaining and expanding code host [integrations](https://docs.sourcegraph.com/integration)_**

Reviewer: ```suggestion
  1. **_Maintaining and expanding [code host integrations](https://docs.sourcegraph.com/dev/background-information/web/code_host_integrations)_**
```

We own front-end code host integrations, but not back-end integrations (the actual connections to code host respositories). Would this doc page be more accurate?

Author: Good call! I'd looked at both of these and had originally wanted to shy away from a /dev/ page (was looking for an overview page) but I think you're right that the /dev/ page is more accurate. For now, I say let's go for it, and I'll note to update this if/when we get a better overview page into the docs.

  1. **_Developing code insights into an entirely new featureset_**

**Outcome**: 

* All of these areas of focus are well-maintained, well-designed, well-documented, well-tested, performant, easy to set up, and leave users impressed. 
* The webapp helps users discover and employ the full power of Sourcegraph effectively, with a high quality and highly usable interface. 
* Soucegraph extensions and the extensions API provide powerful capabilities to users and a great experience for extension developers. 
* Sourcegraph code host native integrations and browser extensions support the most common code hosts and browsers. 
* Code insights expose the value of Sourcegraph's knowledge of your codebase to users at all levels of an organization. 

### Medium term

> _To reach our long-term goal, we set the following medium-term goals to guide our short-term iteration plans. These medium-term goals are listed in order of rough priority. This means we preference a sooner-listed goal when making progress on one would conflict with another goal, but it doesn't mean we only work on the first goal until it's done – we balance our iterations in sum so we can make some progress on all of these goals over a quarter. Within each goal, the sub-goals are also listed in (stricter) priority order._ 

Reviewer 3: ```suggestion
_To reach our long-term goal, we set the following medium-term goals to guide our short-term iteration plans. These medium-term goals are listed in order of rough priority. This means we prefer a sooner-listed goal when making progress on one would conflict with another goal, but it doesn't mean we only work on the first goal until it's done – we balance our iterations in sum so we can make some progress on all of these goals over a quarter. Within each goal, the sub-goals are also listed in (stricter) priority order._ 
```

> _To reach our long-term goal, we set the following medium-term goals to guide our short-term iteration plans. These medium-term goals are listed in order of rought priority. This means we preference the sooner-listed goal when making progress on one would conflict with another goal, but it doesn't mean we only work on the first goal until it's done – we balance our iterations in sum so we can make some progress on all of these goals over a quarter. Within each goal, the sub-goals are also listed in (stricter) priority order._ 

Reviewer 2: ```suggestion
_To reach our long-term goal, we set the following medium-term goals to guide our short-term iteration plans. These medium-term goals are listed in order of rough priority. This means we preference a sooner-listed goal when making progress on one would conflict with another goal, but it doesn't mean we only work on the first goal until it's done – we balance our iterations in sum so we can make some progress on all of these goals over a quarter. Within each goal, the sub-goals are also listed in (stricter) priority order._ 
```

1. **Make the Sourcegraph web interfaces more consistent and improve discoverability of Sourcegraph features.**

>    1. Make progress on design debt throughought the web app where the web team owns the design: repo page areas and panels, navigation and headers, general theming, and the blog page (file viewer). 

Reviewer 5: ```suggestion
   1. Make progress on design debt throughout the web app where the web team owns the design: repo page areas and panels, navigation and headers, general theming, and the blob page (file viewer). 
```

Author: (Thanks for catching! Did that too late at night it seems.)

   1. Build and update designs to help users discover or understand everything Sourcegraph offers, like new navigation menus and header concepts. 

1. **Make Sourcegraph extensions a core part of Sourcegraph users' experiences.**
   1. Track anonymized, general usage of extensions to determine which extensions are most successful at adding value for our users to inform our future work.

>    1. Build new API endpoints that are robust and obviously useful (used immediately in an extension), to grow adoption of extensions. 

Reviewer 3: We mention adding **new** endpoints here, but generally I think removing things to reduce complexity is often just as - if not more - important than adding things. Same goes for renaming, restructuring, changing types, support more return values etc. Can we include that too, so it doesn't sound like the imperative is to add and only add?
We have this captured in the long term goal with "easy to use" etc but are not translating it into medium term goals here.

Author: Sure, I can make this "built, maintain, and update"

   1. Make the extensions action bar clear and scalable, and help users discover and use our extensions. 
   1. Improve discoverability and design of the extension registry to enable users to find the most useful extensions and make users excited to build their own extension. 

1. **Make code insights an entirely new reason to use Sourcegraph.**
   1. Focus on building prototypes to solve the most pressing needs of our customers, to clearly demonstrate the value of code insights. 
   1. Expose more generalizable metrics that let our users measure and track their own goals, whether those are migrations, code smells, security needs, cross-collaboration, or other information about code. 

>    1. Expose features that let our users build their own insights. 

Reviewer 5: To what extent does this goal overlap with the goal of making Sourcegraph more extensible? 

Will the extensions platform have shared ownership once we have separate extensibility and code insights teams, in the same way that the web app is shared between multiple teams? My understanding right now is that these features would be added to the extension API. Maybe this is only a technical/implementation conflict, since code insights and extensions are sufficiently distinct conceptually to create separate teams.

Author: It definitely overlaps some. I think this is more of an eng team organization question (who owns it, where the features go in the codebase) and I will leave that to @jeanduplessis!

If you're also trying to ask "why is this not covered in the extensions goal" I think it's because insights may go beyond extensions. Not all code insights features will necessarily be delivered as extensions – even if they use the extensions API, we might build them into Sourcegraph directly, or build entirely new UX flows around this (rather than "enable the extension in the registry, set it up, etc.). 

Moreover, nothing in the extensions medium-term goal talks about building new extensions, just new endpoints as we need them. We just built some AWESOME extensions, and there's nothing on the brainstorm list of new extensions that is so pressing that I think it's valuable to prioritize that over all these other goals. So even if code insights was just "new extensions that are code insights!" that still wouldn't fit into the medium-term extensions goal now, because "build a new extension, even if it doesn't require a new endpoint" is not part of the extensions goal, but is part of the code insights goal.

Reviewer 2: I can confirm having separate teams for code insights and extensions is purely related to organizational design to allow for better focus and growth. I would expect that a close collaboration between the two teams will be needed. This collaboration will primarily be facilitated by the teams' EMs and PMs.

> 1. **Increase the weekly active users of all our code host integrations.**

Reviewer 3: @Joelkw the code host integrations are basically just implementations of the extension API (other than the web app, which has a lot of other features). How do you think about support of the new extension API functionlities in code host integrations? For example, we would be faced with the decision of whether we implement the directory decorations only for the webapp, or also on code hosts.

Author: Ooh, good question! I can think of some product strategy considerations in this decision, and for now (now = this iteration and the one after it, at least), I'm thinking we leave things specifically like directory decoration out of the code hosts, for the following reasons: 

- I assume the overhead for adding these to code host integrations is "not none" now, and also "not none" in the future. We'd have to build at where to append code within the structure of each code host's directory pages, which means additional overhead that scales linearly with more code hosts. It also means that it could break on a code host if they changed their DOM structure on directory views enough, right? Which means more maintenance overhead as well. 

- I think that the value of our "bring sourcegraph to your code host" integration is mostly around the PR review use case right now, and sometimes in the "googled for a library, found a github/personal website link to a github repo, now I'm browsing the code and sourcegraph.com code intel is cool." In short, around looking at lines of code. I don't think our core value here is "see codecov on the github file view" (and if I'm wrong, we'll hear this from users in feedback like "I want directory decoration on the code host also!" It's a new feature, so we can just truthfully explain we hadn't built it yet - it's not going to break anyone's current workflows). 

- I see advantages to establishing Sourcegraph as the sole place to view code insights, especially as the product is getting started, because it keeps our iteration speed up and feedback loops smaller. At the very least, I see advantages of moving people to "insights viewers" that are not just their code host (Sourcegraph, maybe Backstage, custom BI, anything) because Sourcegraph is best positioned to support UIs built for insights rather than try and fit these insights into an existing page on github. 

Note that these are points specifically for directory decorations. If we were building, say, modals to message someone on slack who wrote the code, those could make more sense to put in the integration. Because this is a case-by-case basis, I think this fits into "Maintain the existing native integrations and browser extensions" because as we build new extensions endpoints, if we decide to put them into the integrations, that is basically now "maintenance" for an existing feature across all our delivery methods.

   1. Maintain the existing native integrations and browser extensions. 
   1. Build support for new code hosts and new browsers, like Safari and Gerrit. 

>    1. Build support for the browser extension to reference both a private Sourcegraph instance and public code on Sourcegraph.com. 

Reviewer 5: What does "more often" mean in this context? Making code host integrations more reliable, increasing user engagement, or both?

Author: I think of "reaching more users" as "total users who've tried an integration" and doing that "more often" is basically my way of saying "they should continue to use the integration, rather than try it (still they are reached) and then stop using it." Basically a different prose-ier way of saying "increase WAU for code host integrations" but given that it's not clear, I'll update that.

1. ✅ Existing sourcegraph extensions are more discoverable ([RFC 209](https://docs.google.com/document/d/1I5BMEGp3QuB81AjSzLCQwq_XJV1sXevlU0lpB4O1pj8/edit#))
1. ✅ The Sourcegraph browser extension is more discoverable and easy to congifure ([RFC 221](https://docs.google.com/document/d/19f4xleYBU1zZZdqMmXlLmFxeR-fwEpOwTOgViOFOnyo/edit))
1. ✅ Build new and improved Sourcegraph extensions to showcase the value and opportunity of extensions ([RFC 246](https://docs.google.com/document/d/1HngEeLNAe7_QzVJr6UPi0Si4ZALqTzb7uonOxUiJP6g/edit))
1. ✅ Improve the Sourcegraph extensions (internal) development experience ([RFC 155](https://docs.google.com/document/d/1ikrUNVe3YVbR-JpegxhjrFdmRkTGzTLcOMkKHnOyjuE/edit)) and (external) documentation
1. 🔄 Code insights migration prototype and directory decoration

> 1. 🔄 Safari browser extension

Reviewer: Is Gerrit support in our short-term plans?

Author: Yep, and glad that you're tracking! I haven't update the short term goals yet beyond these present-time updates ("short term goals on their way!" in the [PR description](https://github.com/sourcegraph/about/pull/1950#issue-515831650)). I wanted to separate the goals from the short term roadmap for the sake of "let's agree we have these goals" before "let's agree this work gets us to these goals." Short term goals PR coming next now.
